src/pipeformer/internal/util.py

```diff
@@ -0,0 +1,192 @@
+__all__ = ("resource_name", "reference_name", "account_arn", "CloudFormationPhysicalResourceCache")
+
+
+def resource_name(resource_type: AWSObject, name: str) -> str:
+    """Build the resource logical name for use in stacks.
+
+    :param resource_type: Resource type
+    :param name: Naive logical name
+    :return: Specific logical name
+    """
+    type_name = resource_type.resource_type.split("::")[-1]
+    return VALUE_SEPARATOR.join((type_name, name))
+
+
+def reference_name(name: str, value_type: str) -> str:
+    """Build the reference name for a resource. Used in stack outputs and parameters.
+
+    :param name: Resource name
+    :param value_type: Value type
+    :return: Specific reference name
+    """
+    return VALUE_SEPARATOR.join((name, value_type))
+
+
+def account_arn(service_prefix: str, resource: str) -> Sub:
+    """Build an IAM policy Arn pattern scoping down as for as possible for the specified service.
+
+    :param service_prefix: Service prefix string
+    :param resource: Any resource data to finish Arn
+    :return: Constructed Sub structure that will resolve to the scoped down Arn
+    """
+    if service_prefix in (awacs.iam.prefix, awacs.s3.prefix):
+        _region = ""
+    else:
+        _region = f"${{{AWS_REGION}}}"
+
+    if service_prefix == awacs.s3.prefix:
+        _account_id = ""
+    else:
+        _account_id = f"${{{AWS_ACCOUNT_ID}}}"
+
+    return Sub(f"arn:${{{AWS_PARTITION}}}:{service_prefix}:{_region}:{_account_id}:{resource}")
+
+
+@attr.s
+class CloudFormationPhysicalResourceCache:
+    """Cache for persistent information about CloudFormation stack resources.
+
+    :param stack_name: Name of target stack
+    :param cache: Pre-populated cache mapping logical resource names to physical resource names (optional)
+    """
+
+    _client = attr.ib()
+    _stack_name: str = attr.ib(validator=instance_of(str))
+    _cache: Dict[str, Dict] = attr.ib(
+        default=attr.Factory(dict),
+        validator=deep_mapping(key_validator=instance_of(str), value_validator=instance_of(str)),
+    )
+
+    def _describe_resource(self, logical_resource_name: str) -> Dict:
+        """Describe the requested resource.
+
+        :param logical_resource_name: Logical resource name of resource to describe
+        :returns: result from ``describe_stack_resource`` call
+        """
+        return self._client.describe_stack_resource(StackName=self._stack_name, LogicalResourceId=logical_resource_name)
+
+    def wait_until_resource_is_complete(self, logical_resource_name: str):
+        """Wait until the specified resource is complete.
+
+        :param logical_resource_name: Logical resource name of resource
+        """
+        response = self.wait_until_resource_exists_in_stack(logical_resource_name)
+        if not response["StackResourceDetail"].get("ResourceStatus", ""):
+            response = self._wait_until_field_exists(logical_resource_name, "ResourceStatus")
+        while True:
```

Comment:
Is tight loop + sleep the best option here? Could we `await` on this instead or something?

Reply:
Created #17 to track this.